src/qibo/derivative.py

```diff
@@ -0,0 +1,117 @@
+import numpy as np
+
+from qibo.config import raise_error
+from qibo.hamiltonians.abstract import AbstractHamiltonian
+
+
+def parameter_shift(
+    circuit,
+    hamiltonian,
+    parameter_index,
+    initial_state=None,
+):
+    """In this method the parameter shift rule (PSR) is implemented.
+    Given a circuit U and an observable H, the PSR allows to calculate the derivative
+    of the expected value of H on the final state with respect to a variational
+    parameter of the circuit.
+
+    Original references:
+        `https://arxiv.org/abs/1811.11184`;
+        `https://arxiv.org/abs/1803.00745`.
+
+    Args:
+        circuit (:class:`qibo.models.circuit.Circuit`): custom quantum circuit
+        hamiltonian (:class: `qibo.hamiltonians.Hamiltonian`): target observable
+        parameter_index (int): the index which identifies the target parameter in the circuit.get_parameters() list
+        initial_state ((1, 2**nqubits) matrix): initial state on which we act with the circuit.
+
+    Returns:
+        np.float value of the derivative of the expected value of H on the final state obtained applying U
+        to the initial state with respect to the target variational parameter.
+
```

Comment:
Suggestion: I think "derivative of the expectation value of the `hamiltonian` with respect to the target variational parameter." should be enough